<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>The Kidney-Genetics Documentation</title>
  <meta name="description" content="This documentation is intended to describe the Kidney-Genetics project" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="The Kidney-Genetics Documentation" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This documentation is intended to describe the Kidney-Genetics project" />
  <meta name="github-repo" content="halbritter-lab/kidney-genetics" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="The Kidney-Genetics Documentation" />
  
  <meta name="twitter:description" content="This documentation is intended to describe the Kidney-Genetics project" />
  

<meta name="author" content="Bernt Popp, Nina Rank, Constantin Wolff, Jan Halbritter" />


<meta name="date" content="2023-10-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="analyses-tables.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.28/datatables.js"></script>
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
<script src="libs/jszip-1.13.4/jszip.min.js"></script>
<script src="libs/pdfmake-1.13.4/pdfmake.js"></script>
<script src="libs/pdfmake-1.13.4/vfs_fonts.js"></script>
<link href="libs/dt-ext-buttons-1.13.4/css/buttons.dataTables.min.css" rel="stylesheet" />
<script src="libs/dt-ext-buttons-1.13.4/js/dataTables.buttons.min.js"></script>
<script src="libs/dt-ext-buttons-1.13.4/js/buttons.html5.min.js"></script>
<script src="libs/dt-ext-buttons-1.13.4/js/buttons.colVis.min.js"></script>
<script src="libs/dt-ext-buttons-1.13.4/js/buttons.print.min.js"></script>
<link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet" />
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script>
<link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet" />
<script src="libs/selectize-0.12.0/selectize.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<script src="libs/plotly-binding-4.10.2/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>



<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Kidney-Genetics Documentation</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#objective"><i class="fa fa-check"></i>Objective</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#methods"><i class="fa fa-check"></i>Methods</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#results"><i class="fa fa-check"></i>Results</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#conclusion"><i class="fa fa-check"></i>Conclusion</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#outlook"><i class="fa fa-check"></i>Outlook</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="analyses-tables.html"><a href="analyses-tables.html"><i class="fa fa-check"></i><b>1</b> Analyses result tables</a>
<ul>
<li class="chapter" data-level="1.1" data-path="analyses-tables.html"><a href="analyses-tables.html#main-table-merged-analyses-sources"><i class="fa fa-check"></i><b>1.1</b> Main table: Merged analyses sources</a></li>
<li class="chapter" data-level="1.2" data-path="analyses-tables.html"><a href="analyses-tables.html#result-table-panelapp"><i class="fa fa-check"></i><b>1.2</b> Result table: PanelApp</a></li>
<li class="chapter" data-level="1.3" data-path="analyses-tables.html"><a href="analyses-tables.html#result-table-literature"><i class="fa fa-check"></i><b>1.3</b> Result table: Literature</a></li>
<li class="chapter" data-level="1.4" data-path="analyses-tables.html"><a href="analyses-tables.html#result-table-diagnostic-panels"><i class="fa fa-check"></i><b>1.4</b> Result table: Diagnostic panels</a></li>
<li class="chapter" data-level="1.5" data-path="analyses-tables.html"><a href="analyses-tables.html#result-table-hpo-in-rare-disease-databases"><i class="fa fa-check"></i><b>1.5</b> Result table: HPO in rare disease databases</a></li>
<li class="chapter" data-level="1.6" data-path="analyses-tables.html"><a href="analyses-tables.html#result-table-pubtator"><i class="fa fa-check"></i><b>1.6</b> Result table: PubTator</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="analyses-plots.html"><a href="analyses-plots.html"><i class="fa fa-check"></i><b>2</b> Analyses plots</a>
<ul>
<li class="chapter" data-level="2.1" data-path="analyses-plots.html"><a href="analyses-plots.html#upset-plot-of-merged-analyses-sources"><i class="fa fa-check"></i><b>2.1</b> UpSet plot of merged analyses sources</a></li>
<li class="chapter" data-level="2.2" data-path="analyses-plots.html"><a href="analyses-plots.html#bar-plot-of-panelapp-results"><i class="fa fa-check"></i><b>2.2</b> Bar plot of PanelApp results</a></li>
<li class="chapter" data-level="2.3" data-path="analyses-plots.html"><a href="analyses-plots.html#bar-plot-of-literature-results"><i class="fa fa-check"></i><b>2.3</b> Bar plot of Literature results</a></li>
<li class="chapter" data-level="2.4" data-path="analyses-plots.html"><a href="analyses-plots.html#bar-plot-of-diagnostic-panels-results"><i class="fa fa-check"></i><b>2.4</b> Bar plot of Diagnostic panels results</a></li>
<li class="chapter" data-level="2.5" data-path="analyses-plots.html"><a href="analyses-plots.html#bar-plot-of-hpo-in-rare-disease-databases-results"><i class="fa fa-check"></i><b>2.5</b> Bar plot of HPO in rare disease databases results</a></li>
<li class="chapter" data-level="2.6" data-path="analyses-plots.html"><a href="analyses-plots.html#bar-plot-of-pubtator-results"><i class="fa fa-check"></i><b>2.6</b> Bar plot of PubTator results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Kidney-Genetics Documentation</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">The Kidney-Genetics Documentation</h1>
<p class="author"><em>Bernt Popp, Nina Rank, Constantin Wolff, Jan Halbritter</em></p>
<p class="date"><em>2023-10-04</em></p>
</div>
<div id="preface" class="section level1 unnumbered hasAnchor">
<h1>Preface<a href="index.html#preface" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<hr />
<p>This documentation is intended to describe the <a href="https://github.com/halbritter-lab/kidney-genetics">Kidney-Genetics</a> project.</p>
<!-- here we place the logo -->
<p><img src="static/img/logo/kidney-genetics_logo.svg" width="25%" style="display: block; margin: auto;" /></p>
<div id="objective" class="section level2 unnumbered hasAnchor">
<h2>Objective<a href="index.html#objective" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<blockquote>
<p>How can we address the lack of a unified and standardized database of kidney disease-associated genes, which hampers diagnosis, treatment, and research comparability in the field of kidney diseases?</p>
</blockquote>
<p>Genetic insights are becoming increasingly influential in the understanding and treatment of various kidney diseases (KD). Hundreds of genes associated with monogenic kidney disease have been identified, providing valuable insights into their diagnosis, management, and monitoring.
However, the lack of a unified and standardized database of genes assigned to kidney diseases has led to diagnostic blind spots and comparability issues among current studies of kidney genetics.
To address this gap, we created the <strong>“Kidney-Genetics”</strong> a regularly updated, automated and publicly accessible database which aims to provide a comprehensive list of all relevant genes associated with kidney disease.</p>
<blockquote>
<p><strong>Key issues:</strong></p>
</blockquote>
<ul>
<li>Create a unified and standardized database of kidney disease-associated genes and provide a valuable resource for the diagnosis, treatment, and monitoring of those diseases<br />
</li>
<li>Allow clinicians and researchers to gain a deeper understanding of the genetic factors underlying different KDs<br />
</li>
<li>Compile, organize and curate important information on the genes to the identify novel candidate genes and genetic variants associated with KDs<br />
</li>
<li>Group and sort the genes into different categories, for example into phenotypic groups, the onset, syndromic, etc.<br />
</li>
<li>Establish genotype-phenotype correlations that can be used to assign multiple clinical entities to a single gene in order to improve understanding and treatment choices<br />
</li>
<li>The information can be used to develop personalized treatment strategies and interventions, leading to more effective and targeted therapies for individuals with KD<br />
</li>
<li>Researchers can freely access “Kidney-Genetics” ensuring consistency and comparability across different research projects, which can accelerate scientific progress, foster collaborations, and facilitate the development of new insights and approaches</li>
</ul>
<p>The scientific literature highlights the need for such a database and emphasizes the importance of genetic research in kidney disease (e.g. <span class="citation">(<a href="#ref-boulogne_kidneynetwork_2023" role="doc-biblioref">Boulogne et al., 2023</a>)</span>).
<!-- TODO: Scan literature that notices a scientific need for a standardized kidney gene database (see some interesting paper in the references section) --></p>
<blockquote>
<p>In summary, our research question and its approach have the potential to provide a deeper scientific understanding of KD genetics, improve diagnostic accuracy, guide treatment selection, advance precision medicine, and facilitate research collaboration. The establishment of the <strong>“Kidney-Genetics”</strong> database addresses an important gap in the field and provides a valuable resource for researchers, clinicians, and patients involved in the discovery and treatment of KD.</p>
</blockquote>
</div>
<div id="methods" class="section level2 unnumbered hasAnchor">
<h2>Methods<a href="index.html#methods" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<blockquote>
<p>To create a thorough and standardized database of kidney-related genes, we employed the following methods and compiled kidney disease-associated gene information from various sources:</p>
</blockquote>
<ol style="list-style-type: decimal">
<li>Utilized data from Genomics England and Australia PanelApp <span class="citation">(<a href="#ref-martin_panelapp_2019" role="doc-biblioref">Martin et al., 2019</a>)</span></li>
<li>Conducted a comprehensive literature review of published gene lists<br />
</li>
<li>Collected information from clinical diagnostic panels for kidney disease<br />
</li>
<li>Performed a Human Phenotype Ontology (HPO)-based <span class="citation">(<a href="#ref-kohler_human_2021" role="doc-biblioref">Köhler et al., 2021</a>)</span>) search in rare disease databases (OMIM)<br />
</li>
<li>Employed a PubTator <span class="citation">(<a href="#ref-wei_pubtator_2013" role="doc-biblioref">Wei et al., 2013</a>)</span> API-based automated literature extraction from PubMed</li>
</ol>
<p>We also developed an evidence-scoring system to differentiate highly confirmed disease genes from candidate genes. We defined the presence of a certain gene in 3 or more of the 5 resources as highly evident genes.
These genes were then manually curated according to predetermined criteria or, in the case of existing ClinGen curation, their data and scores were used. Genes with a score of 2 or less were accordingly more likely to be classified as candidate genes.<br />
Furthermore, we grouped all genes into different categories to later match them in a genotype-phenotype correlation.<br />
<!-- TODO: Provide detailed information about the different groups and categories and explain our decision for them --></p>
<p>To get a more transparent and thus more comprehensive understanding of our several evidence source “pillars”, we listed our different steps below and attached a flowchart for better visualization.</p>
<ol style="list-style-type: decimal">
<li>We retrieved all kidney disease related panels from both PanelApp UK and PanelApp Australia, meaning all panels that include “renal” or “kidney” in its name. That included xxx different lists. The access date was the xxx.<br />
</li>
<li>We identified Genes associated with kidney disease in a systematic Literature search using the following search query:<br />
<em>(1) “Kidney”[Mesh] OR “Kidney Diseases”[Mesh] OR kidney OR renal</em> AND<br />
<em>(2) “Genetic Structures”[Mesh] OR “Genes”[Mesh] OR genetic test OR gene panel OR gene panels OR multigene panel</em> OR targeted panel*<br />
we then screened for published lists and got xxx lists from date to date xxx.<br />
</li>
<li>We used ten common diagnostic panels that can be purchased for genome analysis and extracted the screened genes from them. Those included following panels:
<ul>
<li>Centogene nephrology</li>
<li>Cegat kidney diseases</li>
<li>Preventiongenetics</li>
<li>etc.
<!-- TODO: complete the list and provide the links with RMD --></li>
</ul></li>
<li>We used common databases (e.g. OMIM) for rare diseases and screened them for kidney disease associated Genes from a Human Phenotype Ontology (HPO) based search query.
The most comprehensive HPO term used was “Abnormality of the upper urinary tract” (HP:0010935) and included all subgroup terms. We deliberately chose these to be somewhat broader in order to fully include all relevant kidney diseases such as CAKUT, among others.<br />
</li>
<li>We retrieved all kidney disease associated genes from a PubTator API-based automated literature extraction of publications available on PubMed.</li>
</ol>
<blockquote>
<p><strong>Kidney-Genetics Flowchart</strong></p>
</blockquote>
<div class="figure">
<img src="KidneyGenetics_documentation_files/figure-html/curation_flow_diagram.svg" alt="Curation process flow diagram"  />
<p class="caption">
(#fig:curation_flow_diagram)Curation process flow diagram
</p>
</div>
</div>
<div id="results" class="section level2 unnumbered hasAnchor">
<h2>Results<a href="index.html#results" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<blockquote>
<p>The “Kidney-Genetics” database currently contains detailed information on 2.906 kidney-associated genes with detailed annotations on gene function, kidney phenotype, incidence, possible syndromic disease expression and genetic variation.
To automatically group the genes, we will present the results of phenotypic and functional clustering.<br />
<!-- TODO: Provide clustering and grouping results with numbers and proportions --></p>
</blockquote>
<p>The number of genes extracted from the five analyzed sources of information is as follows: (1) 550, (2) 822, (3) 936, (4) 791, and (5) 2133<br />
Notably, <strong>437</strong> genes (14.6%) of the <strong>total 3001</strong> genes are present in three or more of the analyzed information sources, thus meeting our evidence criteria, indicating high confidence and their potential for diagnostic use.
Of these high evidence genes, <strong>423</strong> (96.8%) are present in at least one, and <strong>56</strong> (12.8%) are present in all 10 comprehensive diagnostic laboratory panels.</p>
<p>To ensure currency, Kidney-Genetics will be updated regularly and automatically at XXX week intervals. We will also provide phenotypic and functional clustering results to facilitate gene grouping.<br />
<!-- TODO: Provide information about the planned update framework --></p>
</div>
<div id="conclusion" class="section level2 unnumbered hasAnchor">
<h2>Conclusion<a href="index.html#conclusion" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<blockquote>
<p>Kidney-Genetics is a comprehensive, free and publicly accessible database that can be used by researchers to analyze genomic data related to KDs. The database will be routinely updated using an automated system and standardized pipeline to ensure that it is always up-to-date with the latest kidney research and diagnostics.</p>
</blockquote>
<p>By utilizing Kidney-Genetics, clinicians, geneticists, and researchers can examine genomic data and improve their understanding of the genetic components of diverse KDs. The code and results are completely available on GitHub. A standardized pipeline and automated system keep our database on the cutting edge of kidney research and diagnostics.
Screening efforts toward manual curation (such as through the ClinGen initiative) and assignment of diagnostic genes to nephrologic disease groups (e.g., syndromic vs. isolated; adult vs. pediatric; cystic, nephrotic, etc.) are currently in the development process and our goals for the near future.</p>
</div>
<div id="outlook" class="section level2 unnumbered hasAnchor">
<h2>Outlook<a href="index.html#outlook" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Future goals include the further manual curation of the high evident genes to acquire a more accurate individual assessment of each gene. For this purpose, we have developed a standardized curation process based on the ClinGen criteria, as previously discussed in the methods section.
Furthermore, diagnostic genes will be assigned to certain defined nephrological disease groups, in order to obtain a phenotype-genotype correlation and gain a better clinical understanding.</p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-boulogne_kidneynetwork_2023" class="csl-entry">
Boulogne, F., Claus, L. R., Wiersma, H., Oelen, R., Schukking, F., Klein, N. de, Li, S., Westra, H.-J., Zwaag, B. van der, Reekum, F. van, Genomics England Research Consortium, Sierks, D., Schönauer, R., Li, Z., Bijlsma, E. K., Bos, W. J. W., Halbritter, J., Knoers, N. V. A. M., Besse, W., … Eerde, A. M. van. (2023). <span>KidneyNetwork</span>: Using kidney-derived gene expression data to predict and prioritize novel genes involved in kidney disease. <em>European Journal of Human Genetics: EJHG</em>. <a href="https://doi.org/10.1038/s41431-023-01296-x">https://doi.org/10.1038/s41431-023-01296-x</a>
</div>
<div id="ref-kohler_human_2021" class="csl-entry">
Köhler, S., Gargano, M., Matentzoglu, N., Carmody, L. C., Lewis-Smith, D., Vasilevsky, N. A., Danis, D., Balagura, G., Baynam, G., Brower, A. M., Callahan, T. J., Chute, C. G., Est, J. L., Galer, P. D., Ganesan, S., Griese, M., Haimel, M., Pazmandi, J., Hanauer, M., … Robinson, P. N. (2021). The <span>Human</span> <span>Phenotype</span> <span>Ontology</span> in 2021. <em>Nucleic Acids Research</em>, <em>49</em>(D1), D1207–D1217. <a href="https://doi.org/10.1093/nar/gkaa1043">https://doi.org/10.1093/nar/gkaa1043</a>
</div>
<div id="ref-martin_panelapp_2019" class="csl-entry">
Martin, A. R., Williams, E., Foulger, R. E., Leigh, S., Daugherty, L. C., Niblock, O., Leong, I. U. S., Smith, K. R., Gerasimenko, O., Haraldsdottir, E., Thomas, E., Scott, R. H., Baple, E., Tucci, A., Brittain, H., De Burca, A., Ibañez, K., Kasperaviciute, D., Smedley, D., … McDonagh, E. M. (2019). <span>PanelApp</span> crowdsources expert knowledge to establish consensus diagnostic gene panels. <em>Nature Genetics</em>, <em>51</em>(11), 1560–1565. <a href="https://doi.org/10.1038/s41588-019-0528-2">https://doi.org/10.1038/s41588-019-0528-2</a>
</div>
<div id="ref-wei_pubtator_2013" class="csl-entry">
Wei, C.-H., Kao, H.-Y., &amp; Lu, Z. (2013). <span>PubTator</span>: A web-based text mining tool for assisting biocuration. <em>Nucleic Acids Research</em>, <em>41</em>(W1), W518–W522. <a href="https://doi.org/10.1093/nar/gkt441">https://doi.org/10.1093/nar/gkt441</a>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="analyses-tables.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/halbritter-lab/kidney-genetics/tree/master/edit_docs/index.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/halbritter-lab/kidney-genetics/tree/master/docs/index.Rmd",
"text": null
},
"download": [["KidneyGenetics_documentation.pdf", "PDF"], ["KidneyGenetics_documentation.epub", "EPUB"], ["KidneyGenetics_documentation.docx", "DOCX"]],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
